Feat/itunes-tunecard UI

DebarshiB13 opened the pull request:
### Ticket Link

---

### Related Links

---

### Description

---
Second PR for the Itunes app.

The new changes are :
Implementation of the UI part.
List of songs displayed from search results in form of Grid
Each song has a card.
Play/Stop songs
Unit tests for each item.

### Steps to Reproduce / Test

---
 * Go to localhost:3000/songs.
 * Type any song or artist name on the search bar.
 * List of cards/ error will appear based on the response.
 * Play/stop any song by clicking the play/stop button.

### GIF's

---
![feat_tunecard_ui](https://user-images.githubusercontent.com/89914287/132449418-0726fdd5-7698-4e3c-9d2d-a4e065ad86b7.gif)
![feat_tunecard_ui_test](https://user-images.githubusercontent.com/89914287/132449439-d457ddad-c814-4334-8bc6-1b84080dbc19.gif)

DebarshiB13 committed:
Created TuneCard Component

Created TuneCard Component to show the individual track with artistname, collectionname
and play/pause button. Added initial test for TuneCard Component.

DebarshiB13 committed:
Added Functionality and Test for TuneCard Component

Added required functionality and related tests for TuneCardComponent.
Updated TuneContainer tests to cover more test scenarios.
Added ids and translation message in en.json.

DebarshiB13 committed:
Used For and If Components

Implemented map of Card using For Component and if checks with If Component

DebarshiB13 committed:
Merge commits from both branches

DebarshiB13 committed:
Added more tests to increase coverage

DebarshiB13 committed:
Fixed merge conflicts

alichherawalla commented on app/components/TuneCard/tests/index.test.js:
```diff
@@ -0,0 +1,59 @@
+
+    const button = getByTestId('play-pause-btn');
+
+    fireEvent.click(button);
```
just so its amply clear can we do this 
```
expect(playStub).not.toHaveBeenCalled();
fireEvent.click(button);
expect(playStub).toHaveBeenCalled();

expect(playStub).not.toHaveBeenCalled();
fireEvent.click(button);
expect(pauseStub).toHaveBeenCalled();
```

DebarshiB13 replied:
Removed the test as the implementation has changed in next commit

alichherawalla commented on app/containers/TuneContainer/index.js:
```diff
@@ -69,14 +89,76 @@ export function TuneContainer({
+        <CustomCard>
+          <Skeleton loading={loading} active>
+            <If condition={searchTerm}>
+              <div>
```
please remove extra nesting

DebarshiB13 committed:
Solved merge conflicts and updated code accordingly

DebarshiB13 committed:
Added Feature To Play Only One Song  At A Time

Added functionality to play one song at a time. Added Tests with 100% test coverage for TuneCard Component.

alichherawalla commented on app/components/TuneCard/index.js:
```diff
@@ -0,0 +1,102 @@
+        data-testid="play-pause-btn"
+        onClick={() => handlePlayPause(previewUrl)}
+        icon={
+          currentTrack?.src === previewUrl ? (
```
use `If` it will read slightly better

DebarshiB13 replied:
Fixed in next commit

alichherawalla commented on app/components/TuneCard/index.js:
```diff
@@ -0,0 +1,102 @@
+  previewUrl,
+  currentTrack,
+  setCurrentTrack,
+  isTrackPlaying,
```
why don't we delegate the playing/pausing of the track to the component itself. Looks like we're controlling it based on some prop values that are passed from outside.

DebarshiB13 committed:
Moved Play/Pause handler to TuneCard Component

Commit:
Added functionality and test for one song play at a time

alichherawalla commented on app/components/TuneCard/index.js:
```diff
@@ -0,0 +1,88 @@
+  const audioRef = useRef();
+  const [isPlaying, setIsPlaying] = useState(false);
+
+  useEffect(() => {
```
this is very inefficient, even if its false your constantly setting it as false. Do we really need to do this? 

we can completely remove the need of the `isPlaying` state right?

DebarshiB13 replied:
Remove the state from TuneCard

alichherawalla commented on app/components/TuneCard/tests/index.test.js:
```diff
@@ -0,0 +1,54 @@
+import TuneCard from '../index';
+
+describe('<TuneCard />', () => {
+  it('should render and match the snapshot', () => {
```
add a test to make sure that everything is artistName, collectioName is getting rendered correctly

DebarshiB13 replied:
Added in next commit

alichherawalla commented on app/containers/TuneContainer/index.js:
```diff
@@ -70,15 +90,87 @@ export function TuneContainer({
+    const resultCount = get(songsData, 'resultCount', 0);
+
+    const handleOnActionClick = (ref, isPlaying) => {
+      // console.log({ ref });
```
remove console

DebarshiB13 replied:
removed

alichherawalla commented on app/containers/TuneContainer/index.js:
```diff
@@ -70,15 +90,87 @@ export function TuneContainer({
+            <If condition={searchTerm} otherwise={null}>
+              <T id="search_term" values={{ searchTerm }} />
+            </If>
+            <If condition={resultCount !== 0}>
```
```
  <If condition={resultCount}>
```
resultCount is truthy

DebarshiB13 replied:
Updated accordingly

alichherawalla commented on app/containers/TuneContainer/index.js:
```diff
@@ -70,15 +90,87 @@ export function TuneContainer({
+                  collectionName={result.collectionName}
+                  cardImg={result.artworkUrl100}
+                  previewUrl={result.previewUrl}
+                  currentTrack={currentTrack}
```
why do we need to pass currentTrack?

DebarshiB13 replied:
removed

DebarshiB13 committed:
Removed isPlaying from TuneCard, added tests for total card items shown etc

DebarshiB13 committed:
Added 100% Test Coverage

DebarshiB13 committed:
Minor Fix

alichherawalla commented on 0001-tests.patch:
```diff
@@ -0,0 +1,140 @@
+From 1217055b0dfd9eda3418496633d92fce76b4dc00 Mon Sep 17 00:00:00 2001
```
please delete the patch file. Add a *.patch rule in the .gitignore as well

DebarshiB13 replied:
Updated accordingly.

DebarshiB13 committed:
Deleted patch file, updated TuneCard index test

DebarshiB13 committed:
Minor Fix

vishnu-wednesday commented on app/components/TuneCard/index.js:
```diff
@@ -0,0 +1,86 @@
+        {artistName}
+      </Typography.Title>
+      <Image src={cardImg} width="100%" preview={false} />
+      <Typography.Paragraph style={{ fontSize: 18 }}>{collectionName}</Typography.Paragraph>
```
A very minor nit. Maybe we can move all these to styled-components and keep that consistency or even use the https://styled-components.com/docs/api#css-prop API

vishnu-wednesday commented on app/containers/TuneContainer/index.js:
```diff
@@ -16,18 +16,37 @@ import tuneContainerSaga from './saga';
 import debounce from 'lodash/debounce';
+import T from '@components/T';
+import { TuneCard } from '@components/TuneCard/index';
+import For from '@components/For/index';
```
`@components/For`

DebarshiB13 replied:
Updated accordingly

vishnu-wednesday commented on app/containers/TuneContainer/index.js:
```diff
@@ -16,18 +16,37 @@ import tuneContainerSaga from './saga';
+import T from '@components/T';
+import { TuneCard } from '@components/TuneCard/index';
+import For from '@components/For/index';
+import If from '@app/components/If/index';
```
Same as above

DebarshiB13 replied:
Updated accordingly

vishnu-wednesday commented on app/containers/TuneContainer/tests/index.test.js:
```diff
@@ -25,25 +26,45 @@ describe('<TuneContainer /> container tests', () => {
-    const songsData = {
-      resultCount: 1,
-      results: [{ id: 1, song: 'Some data' }]
+  it('should render for component when songsData is available', () => {
```
should render `For` component

DebarshiB13 replied:
Updated accordingly

vishnu-wednesday commented on app/containers/TuneContainer/index.js:
```diff
@@ -70,15 +89,82 @@ export function TuneContainer({
+    };
+
+    return (
+      <If condition={results.length !== 0 || loading} otherwise={null}>
```
Do you need the `!==` check?

DebarshiB13 replied:
Updated accordingly

DebarshiB13 committed:
Minor Fix 2

DebarshiB13 committed:
Minor Fix 3, fixed relative imports
